README.md
<p align="center">
    <img src="https://github.com/SeydX/homebridge-bravia-tvos/blob/master/homebridge-ui/ui/src/assets/img/logo.png" height="200">
</p>


# homebridge-bravia-tvos v5

[![npm](https://img.shields.io/npm/v/homebridge-bravia-tvos.svg?style=flat-square)](https://www.npmjs.com/package/homebridge-bravia-tvos)
[![npm](https://img.shields.io/npm/dt/homebridge-bravia-tvos.svg?style=flat-square)](https://www.npmjs.com/package/homebridge-bravia-tvos)
[![GitHub last commit](https://img.shields.io/github/last-commit/SeydX/homebridge-bravia-tvos.svg?style=flat-square)](https://github.com/SeydX/homebridge-bravia-tvos)
[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)
[![Discord](https://img.shields.io/discord/432663330281226270?color=728ED5&logo=discord&label=discord)](https://discord.gg/kqNCe2D)
[![Donate](https://img.shields.io/badge/Donate-PayPal-blue.svg?style=flat-square&maxAge=2592000)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=NP4T3KASWQLD8)

**Creating and maintaining Homebridge plugins consume a lot of time and effort, if you would like to share your appreciation, feel free to "Star" or donate.**

[Click here](https://github.com/SeydX) to review more of my plugins.


## Info

This is a plugin for [Homebridge](https://github.com/nfarina/homebridge) to control your **Sony Bravia Android TV**. 

This plugin supports following functions:

- **Power Switch**
- **Inputs:** like HDMI, Scart, CEC Devices, AV, WIFI etc.
- **Apps:** like YouTube, Prime Video etc.
- **Channels:** Your favourite channels as inputs.
- **Commands:** IRCC commands s inputs.
- **Macros:** A set of IRCC commands as one input
- **Remote control:** native iOS Remote control with customizable commands
- **Authentication:** with Pre-Shared Key or PIN
- **WOL:** supports Wake on Lan
- **Speaker:** with support for two types of speaker (switch or lightbulb)
- **Config UI X Custom UI** 

## Installation instructions

After [Homebridge](https://github.com/nfarina/homebridge) has been installed:

```sudo npm i -g homebridge-bravia-tvos@latest```


## Preparing the TV

### TV Settings

- Set **Remote start** to **ON** _(Settings -> Network -> Remote Start)_
- Change **Authentication** to **Normal and Pre-Shared Key** _(Settings -> Network -> IP Control -> Authentication)_

If you want to use authentication with a Pre-Shared key, please do following steps:

- Enter a **Pre-Shared Key** _(Settings -> Network -> IP control -> Pre-Shared Key)_

## Configuration (Config UI X)

Bravia TV OS v5 supports a custom user interface making configuration via **homebridge-config-ui-x** even easier! Below you can see how easy it is to create, edit or delete a new TV for the config.json using the custom user interface. To use the custom user interface you need at least homebridge-config-ui-x v4.34.0!


https://user-images.githubusercontent.com/34152761/124396673-4b249700-dd0b-11eb-8deb-b7ab18c429e7.mov


## Configuration (Manually)

If you cannot use the custom user interface or want to edit the config.json manually, you must first decide which authentication to use.


### a) PIN Authentication (prefered)

To use the PIN authentication you must first install the bravia module:

``
sudo npm i -g @seydx/bravia@beta
``

And to be able to use the plugin with the PIN procedure your credentials must be created first.

You can create the credentials as follows:

``bravia pair <host> -p 80 -n MyTV``

- \<host\>: The address of your Bravia TV _(eg 192.168.178.99)_
- -n: Name for the app (Default `"@seydx/bravia"`)
- -p: The port of your Bravia TV (Default: `80`)

The PIN displayed on the TV must then be entered in the terminal. This will generate a credentials ``<Object>`` like this:
    
```javascript
{
  name: 'MyTV',
  uuid: 'e9812807-d394-407c-b657-c89a98804e65',
  token: 'A0B9B9D7580466F22EE8F8EA148863774ACCE203',
  expires: 'Fr., 26 Apr. 2009 21:42:48 GMT+00:00'
}
```

Once that is done, you just need to add the "name" to your config.json under "appName". (E.g.: `"appName": "MyTV"`)

**Example:**

```javascript
{
  ...
  "active": true,
  "name": "Sony TV",
  "ip": "192.168.178.123",
  "port": 80,
  "appName": "MyTV",
  ...
}
```

### b) PSK Authentication

- Enter a **Pre-Shared Key** _(Settings -> Network -> IP control -> Pre-Shared Key)_

You need to put the PSK entered in your tv also in your config.json.

**Example:**

```
{
  "name": "Sony TV",
  "ip": "192.168.178.123",
  "psk": "MYPSK"
  ...
}
```

## Example BASIC config

 ```javascript
{
 "bridge": {
   ...
},
 "accessories": [
   ...
],
 "platforms": [
    {
      "platform": "BraviaOSPlatform"
      "name": "BraviaTVOS",
      "debug": true,
      "tvs": [
        {
          "name": "Sony TV",
          "ip": "192.168.178.123",
          "psk": "0000",
          "inputs": [
             {
              "name": "Apple TV",
              "identifier": "Wohnzimmer",
              "source": "cec"
            }
          ],
          "apps": [
            {
              "name": "You Tube",
              "identifier": "YouTube"
            }
          ],
          "channels": [
            {
              "name": "Planet HD",
              "channel": 97,
              "source": "dvbt"
            }
          ],
          "commands": [
            {
              "name": "Volume Up",
              "value": "AAAAAQAAAAEAAABgAw=="
            }
          ],
          "speaker": {
            "active": true,
            "output": "speaker",
            "increaseBy": 1,
            "reduceBy": 1,
            "accType": "lightbulb"
          },
        }
      ]
    }
 ]
}

 ```
 
 See [Example Config](https://github.com/SeydX/homebridge-bravia-tvos/blob/beta/example-config.json) for a **FULL** config example.

 
 ## Options General

| **Attributes** | **Required** | **Usage** | **Default** | **Options** |
|----------------|--------------|-----------|-------------|-------------|
| name | **X** | Name for the log. | `BraviaTVOS`
| debug |  | Enables additional output (debug) in the log. | `false` | `true`, `false`
| warn |  | Enables additional output (warn) in the log. | `true` | `true`, `false`
| error |  | Enables additional output (error) in the log. | `true` | `true`, `false`
| extendedError |  | Enables additional output (detailed error) in the log. | `true` | `true`, `false`

 ## Options TV
 
| **Attributes** | **Required** | **Usage** | **Default** | **Options** |
|----------------|--------------|-----------|-------------|-------------|
| active | **X** | If enabled, the device will be exposed to HomeKit | `false` | `true`, `false`
| ip | **X** | Sony TV IP Address.
| mac | | Sony TV MAC Address.
| port |  | Sony TV Port. | `80`
| psk | **X** | Either psk **OR** appName/appUUID must be setted. (see preparing the TV above)
| appName | **X** | Either psk **OR** appName must be setted. (see preparing the TV above)
| manufacturer | | Manufacturer name for display in the Home app. `Sony`
| model |  | Model name for display in the Home app. | `Bravia`
| serialNumber |  | Serialnumber for display in the Home app. | `00000000`
| refreshInputs |  | When this option is enabled, the TV updates all inputs and saves them to disk for further use. (Please turn on the TV before restarting homebridge with this option enabled, otherwise the plugin will turn on the TV to also retrieve CEC inputs) | `false` | `true`, `false`
| wol |  | When this option is enabled, the plugin uses WOL instead of API to turn on the TV. (WOL must be enabled on the TV and a MAC address must be specified) | `false` | `true`, `false`
| polling |  | Polling interval in seconds. | `10` (s)
| sheduledRefresh |  | Sheduled television (inputs) cache refresh in hours. | `12` (h)


 ## Options TV Inputs
 
| **Attributes** | **Required** | **Usage** | **Default** | **Options** |
|----------------|--------------|-----------|-------------|-------------|
| inputs.name | **X** | Name for the channel to display in the tv inputs list
| inputs.identifier | **X** | Exact name of the input (eg HDMI 1)
| inputs.source | **X** | Type of the tv input |  | `cec`, `component`, `composite`, `hdmi`, `scart`, `widi`

 ## Options TV Apps
 
| **Attributes** | **Required** | **Usage** | **Default** | **Options** |
|----------------|--------------|-----------|-------------|-------------|
| apps.name | **X** | Name for the application to display in the tv inputs list
| apps.identifier | **X** | Exact name of the Application

 ## Options TV Channels
 
| **Attributes** | **Required** | **Usage** | **Default** | **Options** |
|----------------|--------------|-----------|-------------|-------------|
| channels.name | **X** | Name for the channel to display in the tv inputs list
| channels.channel | **X** | Number of the channel as seen on the TV.
| channels.source | **X** | Source of the channel. |  | `dvbt`, `dvbc`, `dvbs`, `isdbt`, `isdbc`, `atsct`, `isdbs`, `analog`

 ## Options TV Commands
 
| **Attributes** | **Required** | **Usage** | **Default** | **Options** |
|----------------|--------------|-----------|-------------|-------------|
| commands.name | **X** | Name for the command to display in the tv inputs list
| commands.value | **X** | IRCC code or name of the command to display in Apple Home. (eg. "AAAAAQAAAAEAAABgAw==" or "PowerOff")

 ## Options TV Macros
 

Changelog.md
# Changelog

# v5.0.8 - 2021-10-08

## Breaking Changes
- Commands with the same values are now merged. If you have used commands as inputs, please remove the TV cache and create a new one by clicking the "refresh" button in the UI and revise your config

## Notable Changes
- Added new tv channel source: `ATSCT`

## Bugfixes
- Minor Bugfixes

# v5.0.7 - 2021-10-02

## Bugfixes
- Fixed an issue where editing tv via config-ui-x displayed a blank screen
- Fixed an issue where `oldModel` TVs would repeatedly display warnings in the log
- Fixed an issue where `sheduledRefresh` was causing an error
- Fixed several issues where channels, inputs and commands were moved to the application cache
- Minor bugfixes

# v5.0.6 - 2021-09-26

## Other Changes
- Support for older TV models which does not support all methods used by this plugin
- Bump dependencies

## Bugfixes
- Fixed an issue where updating the TV via `refreshInputs` saved incorrect data
- Fixed an issue where `oldModel` didnt work properly
- Minor UI bugfixes

# v5.0.5 - 2021-09-21

## Notable Changes
- New Speaker Type: `fan`

## Other Changes
- Refreshing tv cache through config ui x will no longer cached inputs (like hdmi, cec etc.)

## Bugfixes
- Fixed an issue where updating the TV via the config ui x saved incorrect data
- Minor bugfixes

# v5.0.3 - 2021-09-20

## Other Changes
- Refactored `volume selector`: IRCC instead of API calls

## Bugfixes
- Minor Bugfixes

# v5.0.2 - 2021-09-20

## Other Changes
- **Custom UI**
  - Added a `popover` to **Configs** (navi)
- **Homebridge**
  - Added new `oldModel` option to the config.json, which allows to hide errors in the log when the API is not available
- Updated dependencies

## Bugfixes
- Fixed an issue where reloading/fetching inputs didnt work
- Fixed an issue where editing the config with custom ui didnt work
- Fixed an issue where custom remote commands were ignored
- Fixed an issue where changing volume didnt work on some tvs
- Minor Bugfixes
  
# v5.0.1 - 2021-07-06

## Notable Changes
- **Sheduled Refresh**
  - Added new `sheduledRefresh` option to config.json. Sheduled Refresh enables automatic refreshing of the inputs list in the background.

## Other Changes
- **Config**
  - Moved `polling` from **config.polling** to **config.tvs[].polling** to allow specific polling timer for each tv

## Bugfixes
- Fixed an issue where refreshing television (inputs) also removed the television cache (name, input visibility)

# v5.0.0 - 2021-07-04
I am happy to present you version 5 of homebridge-bravia-tvos. The plugin has been redesigned from scratch, offers a new config UI and works much faster than v4. However, most of the changes have been done in the backend to make the plugin even more stable.

Unfortunately some changes had to be made in config.json to improve readability and adapt it to the new version. For this reason I recommend you to delete the TV from your config.json and then reconfigure it via homebridge-config-ui-x. And please take a look at the [example-config.json](https://github.com/SeydX/homebridge-bravia-tvos/blob/master/example-config.json) (If you enter the same name as before as TV name, the TV does not have to be removed from HomeKit).

Have fun with version 5

_- seydx_

---

## Breaking Changes
- **Homebridge**
  - Dropped support for homebridge below v1.3.0
- **Config**
  - **Televisions**
    - Added `active` key to `tvs` to enable/disable the tv without removing it from config. Warning: Default value for active is **false**, this means, that if you DON'T adjust your config.json and add `"active": true` to the tv, the tv will be removed/not exposed from/to HomeKit!
    - Refactored `apps`
    - Refactored `channels`
    - Refactored `commands`
    - Refactored `inputs`
  - **Speaker**
    - Added `active` key to `tvs.speaker` to enable/disable the speaker without removing it from config. Warning: Default value for active is **false**, this means, that if you DON'T adjust your config.json and add `"active": true` to the speaker, the speaker will be removed/not exposed from/to HomeKit!
    - The `Speaker` Accessory will no longer be exposed as an additional accessory in HomeKit. Instead, it now appears as an additional service in the TV.
    - The accessory type `speaker` was removed

## Notable Changes
- **Homebridge UI**
  - The config ui was completely rewritten in `@vue` for a better user experience

## Other Changes
- The code has been refactored (again)
- The backend module `@seydx/bravia` was completely rewritten
- Adjusted config.schema.json to reflect the changes mentioned above
- Updated dependencies

## Bugfixes
- Some minor bugfixes

## v4.1.12 - 2021-05-20

lib/Bravia.js
'use strict';

const axios = require('axios');
const wol = require('wake_on_lan');
const tcpp = require('tcp-ping');
const debug = require('debug')('BraviaPlatformApi');
const base64 = require('base-64');

const tcpping = opt => new Promise((resolve, reject) => tcpp.ping(opt, (err, data) => err ? reject(err) : resolve(data)));
const tcpprobe = (ip,port) => new Promise((resolve, reject) => tcpp.probe(ip,port, (err, available) => err ? reject(err) : resolve(available)));

const timeout = ms => new Promise(res => setTimeout(res, ms));

const isObject = function(a) {
  return (!!a) && (a.constructor === Object);
};

class Bravia {
  constructor(platform, context, waitUntilAuthenticated) {
  
    if(!context.ip) throw 'No ip found in config!';
    
    this.ip = context.ip;
    this.port = context.port || 80;
    this.name = context.name;
    this.context = context;
    
    this.logger = platform.logger;
   
    this.opt = {
      address: this.ip,
      port: this.port,
      timeout: 3000,
      attempts: 500
    };
    
    this.url = 'http://' + this.ip + ':' + this.port;
    
    this.header = {
      'Content-Type': 'application/json'
    };
    
    if(context.psk){
    
      this.logger.info(this.name + ': Initalizing Bravia API with PSK');
    
      this.header['X-Auth-PSK'] = context.psk;
      this.authenticated = true;
      
    } else {
      
      this.HBpath = platform.api.user.storagePath();   
      this.store = require('json-fs-store')(this.HBpath);
      
      if(!waitUntilAuthenticated){
        this.logger.info(this.name + ': Initalizing Bravia API with IP Control');
        this.checkAuthFile();
      } else {
        this.checkUntilAuthenticated();
      }
      
    }
    
    this.count = 0;
    
  }
  
  async checkUntilAuthenticated(){
  
    const self = this;
  
    let name = this.name.includes('Speaker') ? this.name.split(' Speaker')[0] : this.name;
    
    this.store.load(name, async function(err, object){
      if(err) {
      
        //self.logger.error(err);
        
        await timeout(5000);
        self.checkUntilAuthenticated();
      
      } else {

        //check registration
          
        if(object.cookie && object.expires){
  
          const expireDate = new Date(object.expires).getTime();
          let today = new Date();
          today.setDate(today.getDate() + 1);
          today = today.getTime();
    
          if( (expireDate - today) <= 0 ){
    
            await timeout(5000);
            self.checkUntilAuthenticated();
    
          } else {
    
            //everything ok!
            self.header.Cookie = object.cookie;
            self.authenticated = true;
    
          }
  
        } else {
        
          await timeout(5000);
          self.checkUntilAuthenticated();
        
        }
      
      }

    });
  
  }
  
  getAuth(){
    return new Promise((resolve, reject) => {
      this._handleAuth((err, res) => {
        if(err) reject(err);
      
        resolve(res);
      
      });
    });
  }
  
  _handleAuth(callback){
    if(this.authenticated){
      callback(null, 'Authenticated');
    } else {
      this.authCount++;
  
      if(this.authCount >= 12){
  
        callback(this.name + ': Can not authenticate!', null);
        
      } else {
 
        setTimeout(this._handleAuth.bind(this,callback),5000); 
  
      }
    }
  }
  
  async checkAuthFile(){
  
    const self = this;
  
    this.store.load(self.name, async function(err, object){
      if(err) {
      
        let auth = {
          id: self.name,
          pin: 1234
        };
           
        self.store.add(auth, err => {
          if(err) throw err;
          self.checkAuthFile();
        });
      
      } else {

        //check registration
          
        if(object.cookie && object.expires){
  
          const expireDate = new Date(object.expires).getTime();
          let today = new Date();
          today.setDate(today.getDate() + 1);
          today = today.getTime();
    
          if( (expireDate - today) <= 0 ){
    
            //refresh cookie und expire
            self.checkAuthentication(null, null, true);
    
          } else {
    
            //everything ok!
            self.header.Cookie = object.cookie;
            self.authenticated = true;
    
          }
  
        } else {
  
          if(object.pin && (object.pin !== 1234)){
    
            //create cookie and expires
            self.checkAuthentication(object.pin);
    
          } else {
          
            self.checkAuthentication();
    
          }
  
        }
      
      }

    });
  
  }
  
  async checkAuthentication(user, cookie, refresh){
  
    const self = this;
    
    if(!user && !cookie && !refresh)
      this.logger.info(self.name + ': Trying to register on ' + this.ip);
  
    try {
    
      let res = await this.registerCall(user, cookie, refresh);
    
      if(res.data.error && res.data.error.includes(40005)){
          
        //Display off!
        //Turn on display for init auth!
        
        if(!this.wolTried && this.context.mac){
          this.logger.info(self.name + ': Turning on TV for init authentication...');
            
          this.wolTried = true;  
            
          await this.setPowerStatusWOL(this.context.mac);
        } else {
          if(!this.context.mac){
            this.logger.warn(self.name + ': Can not turn on TV via WOL (no mac address in config.json). Please manually turn on the TV!');
          } else {
            this.logger.warn(self.name + ': Can not turn on TV via WOL. Please manually turn on the TV!');
          }
        }
            
        await timeout(15000);
          
        this.checkAuthentication(user, cookie, refresh);
          
      } else {
    
        if(res.headers['set-cookie']){
      
          //cookie received in header!
          //something like this:
    
          /*
           * headers: { 
           *   'content-type': 'application/json',
           *   'content-length': '20',
           *   'connection': 'close',
           *   'set-cookie': [ 
           *     'auth=C78436B5E8CB9A3C6A7DD2B29DE3386A012535A3; 
           *      Path=/sony/; 
           *      Max-Age=1209600; // <- in seconds, need to refresh before it expires!
           *      Expires=Fr., 26 Apr. 2019 21:42:48 GMT+00:00' 
           *   ] 
           * }
          */
          
          let cookieHeader = res.headers['set-cookie'][0].split(';');
          
          cookieHeader.forEach( str => {
  
            if(str.includes('auth')){
              self.header.Cookie = str;
            }
    
            if(str.includes('Expires')){
              self.expires = str.split('=')[1];
            }
          
          });
          
          self.authenticated = true;
    
          this.store.load(self.name, async function(err, object){
            
            let auth = object;
            
            auth.cookie = self.header.Cookie;
            auth.expires = self.expires;
       
            self.store.add(auth, function(err){
            
              if(err) throw err; // err if JSON parsing failed
              
              if(!object.cookie)
                self.logger.info(self.name + ': Authentication successfull!');

              self.logger.info(self.name + ': ' + self.name + '.json refreshed with new cookie!');
              
            });
             
          });
      
        } else {
      
          //no cookie in header!
        
          let error = {
            status: res.status,
            message: res.statusMessage,
            data: res.data,
            headers: res.headers
          };
          
          this.logger.error(JSON.stringify(error));
      
        }
          
      }
      
    } catch(err) {
    
      if(err.response && err.response.status === 401){
      
        //not authorized yet!
        //need to register device
      
        //PIN should be appeared on screen (401)
        //start authentication
      
        this.logger.info(self.name + ': You should see a dialog with a 4-digit PIN.');
        this.logger.info(self.name + ': Please enter the four-digit PIN in ' + self.name + '.json! Located at ' + self.HBpath);
        this.logger.info(self.name + ': Trying again in 30 seconds...');
               
        await timeout(30000);
            
        this.checkAuthFile();
      
      } else {
    
        //something went wrong!
          
        /*let error = {
          code: error.response.status,
          message: error.response.statusMessage,
          data: error.response.data
        };*/
        
        this.logger.error(err);
    
      }
    
    }
  
  }
  
  async registerCall(user, cookie, refresh) {
  
    const self = this;
  
    return new Promise((resolve, reject) => {
    
      let header = {};
    
      if(!refresh){
      
        if(user){
          let basic = 'Basic ' + base64.encode(':' + user);
          header.Authorization = basic;
        }
        
        if(cookie)
          header.Cookie = cookie;
        
      }
      
      var data = '{"method":"actRegister","params":[{"clientid":"BraviaTVOS:481b0a56-fcdf-4b13-bef7-dc22acf965ab","nickname":"BraviaTVOS (homebridge-bravia-tvos)","level":"private"},[{"value":"yes","function":"WOL"}]],"id":8,"version":"1.0"}';
      
      axios({
        url: self.url + '/sony/accessControl',
        method: 'POST',
        data: data,
        headers: header
      }).then(response => {
        resolve(response);
      }).catch(err => {
        reject(err);
      });
    
    });
    
  }
  
  _refreshToken() {
  
    const self = this;

    return new Promise((resolve, reject) => {
    
      if(self.context.psk){
  
        self.header['X-Auth-PSK'] = self.context.psk;
        resolve(true);

      } else {
  
        let name = self.name.includes('Speaker') ? self.name.split(' Speaker')[0] : self.name;

        this.store.load(name, async function(err, object){
    
          if(err) reject(err);

          //check registration
          
          if(object.cookie && object.expires){
  
            const expireDate = new Date(object.expires).getTime();
            let today = new Date();
            today.setDate(today.getDate() + 1);
            today = today.getTime();
    
            if( (expireDate - today) <= 0 ){
    
              //refresh cookie und expire
              self.checkAuthentication(null, null, true);
              resolve(true);
    
            } else {
    
              resolve(true);
    
            }
      
          }

        });

      } 
    });
  }

  async apiCall(url, method, version, params, req) {
  
    const self = this;
  
    let pingError;
  
    if(!await tcpprobe(this.ip,this.port)){
    
      this.logger.warn(this.name + ': TV currently offline, trying to connect (this can take up to 5 minutes) ...');
    
      try {
      
        await tcpping(this.opt);
        await timeout(10000);
        
        this.logger.info(this.name + ': Connection established!');
      
      } catch(err){
      
        pingError = true;
      
      }
      
    }    
  
    req = req ? req : 'post';   
    params = params ? [params] : [];
    
    this.count++;
    let currCount = this.count;
    debug('[Bravia Debug]: api request ' + currCount + ': ' + req + ' ' + url + ' - ' + method + (params.length ? ' - ' + JSON.stringify(params) : ''));
 
    return new Promise((resolve, reject) => {
      if(self.url && (self.header['X-Auth-PSK'] || self.header.Cookie) && !pingError){
      
        this._refreshToken().then(() => {
    
          let data;
          let header = this.header;
        
          if(url === '/sony/IRCC'){
    
            data = '<?xml version="1.0"?><s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body><u:X_SendIRCC xmlns:u="urn:schemas-sony-com:service:IRCC:1"><IRCCCode>' + method + '</IRCCCode></u:X_SendIRCC></s:Body></s:Envelope>';

            header['SOAPACTION'] = '"urn:schemas-sony-com:service:IRCC:1#X_SendIRCC"';
            header['Content-Type'] = 'text/xml';
            header['Content-Length'] = Buffer.byteLength(data);
    
          } else {
    
            header = this.header;
    
            data = {
              'id':1,
              'method':method,
              'version':version,
              'params':params
            };
    
            data = JSON.stringify(data);
    
          }
      
          axios({
            url: self.url + url,
            method: req,
            data: data,
            headers: header
          }).then(response => {
    
            let result;
    
            if(response.data.result){
    
              if(response.data.result.length){
                result = response.data.result[0];
              } else {
                result = 'Ok';
              }
            
              debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + JSON.stringify(result)); 
    
            } else if(response.data.results){
          
              if(response.data.results.length){
                result = response.data.results[0];
              } else {
                result = 'Ok';
              }
            
              debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + JSON.stringify(result)); 
          
            } else if(response.data.error) {
          
              if(response.data.error.includes('Illegal State')){
        
                result = 'App State';
        
                debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + result);
      
              } else if(response.data.error.includes('Display Is Turned off')){
        
                result = 'Display Off';
        
                debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + result);
        
              } else {
        
                debug('[Bravia Debug]: api request ' + currCount + ': Error - ' + JSON.stringify(response.data.error)); 
    
                reject(response.data.error); 
        
              }
    
            } else {
          
              if((isObject(response.data)||typeof response.data === 'string'||response.data instanceof String) && response.data.includes('<?xml version="1.0"?>')){
        
                result = 'Ok';
        
                debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + JSON.stringify(response.data).replace(/\s/g, ''));
        
              } else {
            
                let error = {
                  status: response.status,
                  message: response.statusText,
                  config: response.config,
                  data: response.data
                };
        
                debug('[Bravia Debug]: api request ' + currCount + ': Error Unknown');
    
                reject(error);
        
              }
    
            }
    
            resolve(result);
          }).catch(err => {
          
            let error = err;
        
            if(err.response){
              error = {
                status: err.response.status,
                message: err.response.statusText,
                config: err.config,
                data: err.response.data
              };
            } 
            
            debug('[Bravia Debug]: api request ' + currCount + ': Error Unknown');  
        
            reject(error);
          });
        
        
        }).catch(err => {
          debug('[Bravia Debug]: api request ' + currCount + ': Error!');
          reject(err);  
        });
        
      } else {
        if(!self.url){
          reject('No URL');
        } else if(!self.header['X-Auth-PSK'] || !self.header.Cookie){
          reject('No PSK/Cookie!');
        } else {
          reject('Can not connect to TV!');
        }
      }
    });
  }
  
  wolCall(mac){
  
    return new Promise((resolve, reject) => {
    
      debug('[Bravia Debug]: WOL: send magic packets to ' + mac);
    
      wol.wake(mac, function(error) {
        if (error) {
          debug('[Bravia Debug]: WOL: Error: ' + JSON.stringify(error));
          reject(JSON.stringify(error));
        } else {
          debug('[Bravia Debug]: WOL: Ok');
          resolve('Done');
        }
      });
  
    });  
  
  }
  
  //# /sony/guide
  
  getServiceProtocols(){
    return this.apiCall('/sony/guide', 'getServiceProtocols', '1.0');
  }
  
  getSupportedApiInfo(){
    return this.apiCall('/sony/guide', 'getSupportedApiInfo', '1.0');
  }

  //# /sony/system
  
  getColorKeysLayout() {
    return this.apiCall('/sony/system', 'getColorKeysLayout', '1.0');
  }
  
  getLEDIndicatorStatus() {
    return this.apiCall('/sony/system', 'getLEDIndicatorStatus', '1.0');
  }
  
  getCurrentTime() {
    return this.apiCall('/sony/system', 'getCurrentTime', '1.0');
  }
  
  getDateTimeFormat() {
    return this.apiCall('/sony/system', 'getDateTimeFormat', '1.0');
  }
  
  getNetworkSettings(type) {
    return this.apiCall('/sony/system', 'getNetworkSettings', '1.0', {'netif': type});
  }
  
  getPowerSavingMode() {
    return this.apiCall('/sony/system', 'getPowerSavingMode', '1.0');
  }
  
  getPowerStatus() {
    return this.apiCall('/sony/system', 'getPowerStatus', '1.0');
  }
  
  getRemoteControllerInfo() {
    return this.apiCall('/sony/system', 'getRemoteControllerInfo', '1.0');
  }
  
  getSystemInformation() {
    return this.apiCall('/sony/system', 'getSystemInformation', '1.0');
  }
  
  getInterfaceInformation() {
    return this.apiCall('/sony/system', 'getInterfaceInformation', '1.0');
  }
  
  getSystemSupportedFunction() {
    return this.apiCall('/sony/system', 'getSystemSupportedFunction', '1.0');
  }
  
  getWolMode() {
    return this.apiCall('/sony/system', 'getWolMode', '1.0');
  }
  
  setWolMode(state) {
    if(state===undefined) return 'No state defined!';
    return this.apiCall('/sony/system', 'setWolMode', '1.0', {enabled: state});
  }
  
  setPowerStatus(state) {
    if(state===undefined) return 'No state defined!';
    return this.apiCall('/sony/system', 'setPowerStatus', '1.0', {status: state});
  }
  
  setPowerStatusWOL(mac) {
    if(mac===undefined) return 'No mac address defined!';
    return this.wolCall(mac);
  }
  
  //# /sony/videoScreen
  
  getAudioSourceScreen() {
    return this.apiCall('/sony/videoScreen', 'getAudioSourceScreen', '1.0');
  }
  
  getPipSubScreenPosition() {
    return this.apiCall('/sony/videoScreen', 'getPipSubScreenPosition', '1.0');
  }
  
  //# /sony/audio
  
  getVolumeInformation() {
    return this.apiCall('/sony/audio', 'getVolumeInformation', '1.0', '1.1');
  }
  
  setAudioMute(state) {
    if(state===undefined) return 'No state defined!';
    return this.apiCall('/sony/audio', 'setAudioMute', '1.0', {status: state});
  }
  
  // target = headphone|speaker - volume: 1-100
  setAudioVolume(target, volume) {
    target = target === undefined ? 'speaker' : target;
    volume = volume === undefined ? 10 : volume;
    return this.apiCall('/sony/audio', 'setAudioVolume', '1.0', {target: target, volume: volume.toString()});
  }
  
  //# /sony/avContent
  
  // example: source = extInput:hdmi
  getContentCount(source) {
    if(source===undefined) return 'No source defined!';
    return this.apiCall('/sony/avContent', 'getContentCount', '1.1', {source: source});
  }
  
  // source = usb:recStorage|tv:dvbc,tv:dvbt
  getContentList(source, cnt, stIdx) {
    if(source===undefined) return 'No source defined!';
    return this.apiCall('/sony/avContent', 'getContentList', '1.2', {source: source, cnt: cnt||200, stIdx: stIdx||0});
  }
  
  getCurrentExternalInputsStatus() {
    return this.apiCall('/sony/avContent', 'getCurrentExternalInputsStatus', '1.0');
  }
  
  getParentalRatingSettings() {
    return this.apiCall('/sony/avContent', 'getParentalRatingSettings', '1.0');
  }
  
  getPlayingContentInfo() {
    return this.apiCall('/sony/avContent', 'getPlayingContentInfo', '1.0');
  }
  
  getSchemeList() {
    return this.apiCall('/sony/avContent', 'getSchemeList', '1.0');
  }
  
  // scheme = exInput|fav|tv|usb
  getSourceList(scheme) {
    if(scheme===undefined) return 'No scheme defined';
    return this.apiCall('/sony/avContent', 'getSourceList', '1.0', {scheme: scheme});
  }
  
  // e.g uri = extInput:hdmi?port=1
  setPlayContent(uri) {
    if(uri===undefined) return 'No uri defined';
    return this.apiCall('/sony/avContent', 'setPlayContent', '1.0', {uri: uri});
  }
  
  //# /sony/recording
  
  getRecordingStatus() {
    return this.apiCall('/sony/recording', 'getRecordingStatus', '1.0');
  }
  
  getSupportedRepeatType() {
    return this.apiCall('/sony/recording', 'getSupportedRepeatType', '1.0');
  }
  
  //# /sony/appControl
  
  getApplicationList() {
    return this.apiCall('/sony/appControl', 'getApplicationList', '1.0');
  }
  
  getApplicationStatusList() {
    return this.apiCall('/sony/appControl', 'getApplicationStatusList', '1.0');
  }
  
  setActiveApp(uri) {
    if(uri===undefined) return 'No uri defined';
    return this.apiCall('/sony/appControl', 'setActiveApp', '1.0', {uri: uri});
  }
  
  terminateApps() {
    return this.apiCall('/sony/appControl', 'terminateApps', '1.0');
  }
  
  //# endpoints available with 'getServiceProtocols'
  
  getMethodTypes(endpoint, versions){
    if(endpoint===undefined) return 'Endpoint required!';
    if(versions===undefined) versions = '1.0';
    return this.apiCall(`/sony/${endpoint}`, 'getMethodTypes', '1.0', versions);
  }
  
  getVersions(endpoint){
    if(endpoint===undefined) return 'Endpoint required!';
    return this.apiCall(`/sony/${endpoint}`, 'getVersions', '1.0');
  }
  
  //# IRCC
  
  setIRCC(command) {
    if(command===undefined) return 'No command defined!';
    return this.apiCall('/sony/IRCC', command);
  }
  
}

module.exports = Bravia;
